src/blog/tanstack-router-route-matching-tree-rewrite.md

```diff
@@ -0,0 +1,230 @@
+---
+title: How we accidentally made route matching more performant by aiming for correctness
+published: 2025-11-18
+authors:
+  - Florian Pellet
+---
+
+![Big performance number](/blog-assets/tanstack-router-route-matching-tree-rewrite/header.png)
+
+We achieved a 20,000× performance improvement in route matching in TanStack Router. Let's be honest, this is _definitely_ cherry-picked, but the number is real and comes from a real production application. More importantly, it shows that matching a pathname to a route is no longer bottlenecked by the number of routes in your application.
+
+## The Real Problem: correctness, not speed
+
+One big responsibility of a router is to match a given URL pathname (e.g., `/users/123`) to a route definition (e.g., `/users/$userId`). This is deceptively complex when you consider all the different types of route segments (static, dynamic, optional, wildcard) and the priority rules that govern which route should match first.
+
+Our previous route matching algorithm would look through every route in the route tree, and through a mix of pattern matching, manual look-aheads, and recursion, find the best match. As we added more features like optional segments and wildcards, the algorithm became increasingly complex and slow, and we started receiving reports of incorrect matches.
```

Comment:
can we bring this in:

base of the previous algorithm was trying to sort all routes into a flat list.
since our sorting did not adhere to a strict weak ordering, sorting was not well defined and even behaved differently between Chrome and Firefox